jdi-light/src/main/java/com/epam/jdi/light/elements/base/JDIBase2.java

```diff
@@ -0,0 +1,525 @@
+import com.epam.jdi.light.elements.composite.WebPage;
+import com.epam.jdi.light.elements.interfaces.base.*;
+import com.epam.jdi.tools.*;
+import com.epam.jdi.tools.func.*;
+import com.epam.jdi.tools.map.MapArray;
+import org.openqa.selenium.*;
+import org.openqa.selenium.interactions.Actions;
+
+import java.util.List;
+
+import static com.epam.jdi.light.common.Exceptions.*;
+import static com.epam.jdi.light.common.SearchStrategies.*;
+import static com.epam.jdi.light.driver.WebDriverByUtils.*;
+import static com.epam.jdi.light.elements.base.OutputTemplates.*;
+import static com.epam.jdi.light.elements.init.UIFactory.*;
+import static com.epam.jdi.light.logger.LogLevels.*;
+import static com.epam.jdi.light.settings.JDISettings.*;
+import static com.epam.jdi.light.settings.WebSettings.*;
+import static com.epam.jdi.tools.EnumUtils.*;
+import static com.epam.jdi.tools.LinqUtils.*;
+import static com.epam.jdi.tools.ReflectionUtils.*;
+import static com.epam.jdi.tools.StringUtils.*;
+import static com.epam.jdi.tools.switcher.SwitchActions.*;
+import static java.util.Collections.*;
+import static org.apache.commons.lang3.StringUtils.*;
+
+/**
+ * Created by Roman Iovlev on 14.02.2018
+ * Email: roman.iovlev.jdi@gmail.com; Skype: roman.iovlev
+ */
+public abstract class JDIBase extends DriverBase implements IBaseElement, HasCache {
+    public static JFunc1<String, String> STRING_SIMPLIFY =
+        s -> s.toLowerCase().replaceAll("[^a-zA-Z0-9]", "");
+
+    public JDIBase base() {
+        return this;
+    }
+    public JDIBase() {
+      // searchRules.add(ELEMENT.searchRule);
+    }
+    public JDIBase(JDIBase base) {
+        setCore(base);
+    }
+    public JDIBase setCore(JDIBase base) {
+        locator = base.locator.copy();
+        name = base.name;
+        parent = base.parent;
+        varName = base.varName;
+        typeName = base.typeName;
+        failElement = base.failElement;
+        driverName = base.driverName;
+        context = base.printFullLocator();
+        webElement = base.webElement.copy();
+        webElements = base.webElements.copy();
+        searchRules = base.searchRules.copy();
+        beforeSearch = base.beforeSearch;
+        timeout = base.timeout;
+        return this;
+    }
+    public MapArray<String, Object> params = new MapArray<>();
+
+    public JDILocator locator = new JDILocator(this);
+    @Override
+    public DriverBase setParent(Object parent) {
+        //this.locator.isRoot = false;
+        return super.setParent(parent);
+    }
+    public CacheValue<WebElement> webElement = new CacheValue<>();
+    public CacheValue<List<WebElement>> webElements = new CacheValue<>();
+    public MapArray<String, JFunc1<WebElement, Boolean>> searchRules = new MapArray<>();
+    private MapArray<String, JFunc1<WebElement, Boolean>> searchRules() {
+        return searchRules;
+    }
+    public JAction1<UIElement> beforeSearch = null;
+    WebElement beforeSearch(WebElement el) {
```

Comment:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Issue found: [Use explicit scoping instead of the default package private level](https://app.codacy.com/manual/42_/jdi-light/pullRequest?prid=5327794)